src/java.base/share/classes/java/lang/Object.java

```diff
@@ -95,6 +95,11 @@ public Object() {}
      * As far as is reasonably practical, the {@code hashCode} method defined
      * by class {@code Object} returns distinct integers for distinct objects.
      *
+     * <p>This method should <em>not</em> throw any exceptions. The
+     * implementation should not use excessive memory or time for its
+     * computations and should return a result for cyclic data
+     * structures.
```

Comment:
Hello Joe, adding this text to set the expectations looks reasonable. However, I think the text "should return a result for cyclic data structures." feels a bit odd. If I understand correctly what it's trying to state is that for a class of the form:

```java
class Foo {
   Bar bar;
   Foo self;
   
   public void setSelf() {
      this.self = this;
   }
}
```
then:

```
Foo foo = new Foo();
foo.toString(); // or foo.hashCode()
```
should be able to return the output from hashCode() and toString() even when an instance of `Foo` has a field which holds the same `Foo` instance. i.e. `toString()` and `hashCode()` should be able to handle re-entrancy and shouldn't consume excessive memory or time.

If that understanding is correct, then maybe we could improve that text to state it differently instead of cyclic data structures? I checked the JDK repo to see if this term has previously been used but it hasn't. Maybe we should just state that the hashCode() toString() methods should be able to deal with re-entrancy?

Reply:
My understanding is the same. Personally, the wording reads unambiguously. The words "circular" and "circularity" can be seen throughout codebase, in similar contexts.

FWIW, a simpler example can be constructed like this:

    var list = new ArrayList<>(); 
    list.add(list);

Reply:
Changed to "cyclic/circular".

The "A" in "DAG" stands for "acyclic", but circular may be more commonly used.

Reply:
While I understand and agree with the intent, we have several exceptions in the OpenJDK itself.

Circular structures that throw `StackOverflowError` on `hashCode()` and huge numbers that take 20-30 min to compute `toString()` are some examples.

Reply:
> While I understand and agree with the intent, we have several exceptions in the OpenJDK itself.
> 
> Circular structures that throw `StackOverflowError` on `hashCode()` and huge numbers that take 20-30 min to compute `toString()` are some examples.

I wouldn't be surprised if there are some cases like that in the JDK. However, if those types are generally usable outside of the JDK, I would advocate for the implementations being updated to behave better.

I added the cyclic/circular guidance in part based on work I did add suppressed exceptions to Throwable in JDK 7. One tricky part of the implementation was making sure the stack trace output did cycle detection since there were various ways exceptions could end up referring to each other.

Reply:
> One tricky part of the implementation was making sure the stack trace output did cycle detection since there were various ways exceptions could end up referring to each other.

Yes, the tricky part was to "Guard against malicious overrides of Throwable.equals".

Reply:
> My understanding is the same. Personally, the wording reads unambiguously. The words "circular" and "circularity" can be seen throughout codebase, in similar contexts.
> 
> FWIW, a simpler example can be constructed like this:
> 
> ```
> var list = new ArrayList<>(); 
> list.add(list);
> ```

PS There are cases where cyclic structures are not supported, such as in java.util.Set:

" A special case of this prohibition is that it is not permissible for a set to contain itself as an element."

Reply:
> PS There are cases where cyclic structures are not supported, such as in java.util.Set:

There are algorithms that also don't support that:

    /**
     * Returns a hash code based on the "deep contents" of the specified
     * array.  If the array contains other arrays as elements, the
     * hash code is based on their contents and so on, ad infinitum.
     * It is therefore unacceptable to invoke this method on an array that
     * contains itself as an element, either directly or indirectly through
     * one or more levels of arrays.  The behavior of such an invocation is
     * undefined.
    
     ...

     */
    public static int deepHashCode(Object[] a) {